### ISSUE TITLE
How to pass down tracing information in ServeHTTP to long-running tasks?

### ISSUE THREAD

**Issue by fatih** [census-instrumentation/opencensus-go#987]
Hi,

I have a process that is triggered via a webhook (ServeHTTP) that I wanted to trace. In my main package I use this simple setup:

```
func main() {
	handler = &ochttp.Handler{
		Propagation: &propagation.HTTPFormat{},
		Handler:     &MyHandler{},
	}

	srv := &http.Server{
		Addr:         "0.0.0.0:" + port,
		WriteTimeout: time.Second * 15,
		ReadTimeout:  time.Second * 15,
		IdleTimeout:  time.Second * 60,
		Handler:      handler,
	}

	srv.ListenAndServe()
}
```

And in my handler my setup is like this:

```
func (m *MyHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// this foo() function takes some time to run and doesn't depend on writing
	// anything to http.ResponseWriter, so we can process it in the background,
	// it takes approximately around 3-15 seconds depending on the incoming
	// information.  I want to debug it via tracing, to see why it takes
	// sometimes 3 seconds, but in some cases 15 seconds :)

	// this doesn't work because ctx is cancelled immediately when I return :(
	// go foo(cxt)

	// I have to pass a new context, but then I lose the tracing information
	// injected from ochttp.Handler (i.e:
	// https://github.com/census-instrumentation/opencensus-go/blob/master/plugin/ochttp/server.go#L137)
	go foo(context.Background())
}

func foo(ctx context.Context) {
	ctx, span := trace.StartSpan(ctx, "foo")
	defer span.End()

	bar(ctx)
}

func bar(ctx context.Context) {
	ctx, span := trace.StartSpan(ctx, "bar")
	defer span.End()
}
```

The problem I have is, I can't pass down the `r.Context()`, because it's cancelled immediately when I return from `ServeHTTP`. According to the doc:

```
// For incoming server requests, the context is canceled when the
// client's connection closes, the request is canceled (with HTTP/2),
// or when the ServeHTTP method returns.
func (r *Request) Context() context.Context {
	if r.ctx != nil {
		return r.ctx
	}
	return context.Background()
}
```

I therefore can't associate the incoming request with my long process `foo()` and lose all the information that is injected via `ochttp.Handler`, see: https://github.com/census-instrumentation/opencensus-go/blob/master/plugin/ochttp/server.go#L137 

I'm able to trace `foo()` and all the child functions associated with it when I use context.Background() (and therefore don't have any cancellation attached to it), but this means than that I can't associate anymore with the incoming request

I'm fairly new using tracing with `context.Context` and I'm wondering what the idiomatic way of solving this is?  Feel free to close if this is the wrong place.

Thanks

**Comment by odeke-em** [comment]
Hello @fatih, great to catch you here, thank you for the question and welcome to the OpenCensus project!


So this problem can be handled in a couple of ways:
a) Using [trace.FromContext](https://godoc.org/go.opencensus.io/trace#FromContext) to retrieve the span from r.Context()
```go
spanA := trace.FromContext(ctx)
// Then use spanA
```
but that unfortunately adding children that might start after the parent has ended isn't great. Please read onto the next suggestion

b) Because the span is from the HTTP request i.e. the rest of the work will be processed asynchronously and due to Go runtime scheduling (hose goroutines could be scheduled even after the function has returned since there isn't explicit synchronization).
This would be a good usecase for a Span.Link as documented on the website https://opencensus.io/tracing/span/link/#uses-and-advisory and with code samples https://opencensus.io/tracing/span/link/#0
```go
func handle(w http.ResponseWriter, r *http.Request) {
   ctx := r.Context()
   spawningSpan := trace.FromContext(ctx)

   go foo(context.Background, spawningSpan)
}

func foo(ctx context.Context, spawningSpan *trace.Span) {
    ctx, span := trace.StartSpan(ctx, "foo")
   defer span.End()

   addAsLink(span, spawningSpan)
   bar(ctx)
}

func bar(ctx context.Context) {
    ctx, span := trace.StartSpan(ctx, "bar")
    defer span.End()
}

func addAsLink(subjectSpan, spawningSpan *trace.Span) bool {
   if spawningSpan == nil {
      return false
   }
   subjectSpan.AddLink(trace.Link{
        TraceID: spawningSpan.TraceID,
        SpanID:  spawningSpan.SpanID,
        Type:    trace.LinkTypeParent,
        Attributes: map[string]interface{}{
                "reason": "webhookSpan",
        },
   })
   return true
}
```

Which will allow you to later correlate traces. Some backends display links or at least allow you to click on a Span.Link but anyways, any asynchronous processing can be alleviated by using Span.AddLink.

I'll also page some more collaborators to voice some more ideas if they'd like @bogdandrutu @rakyll 

Hope this helps and please feel free to ask more questions and submit any bugs.
Thank you and hope this helps.

**Comment by fatih** [comment]
Thanks @odeke-em for the great answer. 

You said ` adding children that might start after the parent has ended isn't great` and I totally agree with it. What if we create a new parent ourself in ServeHTTP? I also saw this link (https://github.com/golang/go/issues/28279#issuecomment-431613687), and someone suggested this form:

ctx = trace.NewContext(tag.NewContext(context.Background, tag.FromContext(ctx)), trace.FromContext(ctx))

I believe this would create a new Parent, or not ? 

2. I believe linking looks really useful, let me experiment with it.

**Comment by basvanbeek** [comment]
> @basvanbeek I'm using StackDriver.
> 
> So do you mean, that
> 
> ```
> ctx = trace.NewContext(tag.NewContext(context.Background, tag.FromContext(ctx)), trace.FromContext(ctx))
> ```
> will create a new Parent and then the span in `foo()` will become a child, without linking?
> 
Let me rephrase for clarity. A parent is not created, it can become one if a new span refers to it as being its parent. This particular piece of code just copies over the tag values and the already existing span from one context to the other without changing anything with respect to your trace.

> I'm not against linking btw, I just want to understand what's going on before I blindly go and implementing via linking :) So, linking is superior than creating a new parent and inject the values from the Grandparent ?

Linking is an additional concept allowing for correlating relationships between different traces or within traces where spans do not follow parent-child relationship. 

In your particular case I personally see no issue with a parent-child relationship as you expect lifetimes of parent and child to not follow RPC boundaries (child lifetime being within parent's lifetime). Visually speaking if your child will outlive its parent and potentially spawn after parent deceases it will show up nicely in the trace timeline and be quite clear as to what happened.

This does become an issue if your idea of a trace should not outlive root request/response and trace should be request/response scoped critical path only. Then links come into play.

**Comment by fatih** [comment]
Thanks a lot again @odeke-em and @basvanbeek. I successfully integrated OC into my project by pulling out the span and tags via the `FromContext()` functions and then passing it to my child functions. Everything works excellent so far.

**Comment by bogdandrutu** [comment]
@fatih can you add a short code snippet of what you did? I am very interesting in this topic.
